### ISSUE TITLE
Possible bug when using a dispatcher

### ISSUE THREAD

**Issue by mikemm27** [jmoenig/Snap--Build-Your-Own-Blocks#891]
I probably found a bug that I tried to make evident with the attached Snap! code. I shortly describe it.
In the first part a list of two method bodies is initialized (variable ml). These two similar methods say something and then turn the current sprite. The ml list is passed to a dispatcher (block 'make an object'). 
Then the 'check' block is run on the Sprite2 sprite. Correctly the first commands of this block are (apparently) executed by Sprite2, included the indirect call of the 'check2' block through the rep variable. But when the dispatcher is called with the m1 method, the say and the turn commands are apparently run on the first sprite (Sprite) which is the sprite associated with the initial 'when green flag clicked' code. The sequence of these two commands is repeated and again the code activated by the dispatcher, and only this, is run on the first sprite. 

I thank you in advance if you would analyze this case and correct the bug or explain my possible misunderstanding.
mike

![bug01](https://cloud.githubusercontent.com/assets/11302374/8894464/ac8dbf2a-33b7-11e5-8f95-8932317026c7.png)
![bug02](https://cloud.githubusercontent.com/assets/11302374/8894471/b439ffc2-33b7-11e5-9664-27ff6d2afd7d.png)
![bug03](https://cloud.githubusercontent.com/assets/11302374/8894473/ba698db8-33b7-11e5-950c-c6b0dbfe4262.png)
![bug04](https://cloud.githubusercontent.com/assets/11302374/8894475/c0112bb8-33b7-11e5-8621-663b0cb1ded9.png)

**Comment by jmoenig** [comment]
Not exactly sure whether I understand the issue here. Could you perhaps post a link to a shared project? From your description it sounds to me that what Snap is doing is intentional. Snap's *ring* bind their environment - including the sprite in whose context they will execute - when they're *reified*, not when they're called. If you want to run a script or reporter in the context of another Sprite you can use the ```of``` block for this. Does this make sense?

**Comment by mikemm27** [comment]
What happens if I attach the project to this replying message?

As you can see, I used the 'of' block at the end of the green flag code,
so it is supposed that from this moment and until the check block
returns, all commands are linked to Sprite2. But indeed this works
for all the commands in the check block apart from  the one
with the couple of nested call (the dispatcher) when say and
move/turn commands are executed ON the sprite which was active in the
first part of the green flag code. I copied a similar code also
as the Sprite2(2) sprite's script and if you execute it typing 'space',
now the dispatcher is executed on this sprite (Sprite2(2)).
I hope it is clear now.
Thank you for your attention.
mike


Jens Mönig <notifications@github.com> ha scritto:

**Comment by jmoenig** [comment]
You can save the project to the SnapCloud and share it. If you then open it again, you can copy and paste the url in your browser here. Thanks!

**Comment by mikemm27** [comment]
I tried both yesterday and today to sign up to the SnapCloud but the site seems unreachable (see below). I would prefer not to use a personal rep. (dropbox or similar) providing te link here: may I send you the project xml somewhere else?
mike

![schermata 07-2457232 alle 12 16 36](https://cloud.githubusercontent.com/assets/11302374/8928943/222b499e-3522-11e5-99c1-4b5ef96cf2d4.png)

**Comment by mikemm27** [comment]
Ok, I eventually succeeded in signup, upload and share: the link is

http://snap.berkeley.edu/snapsource/snap.html#present:Username=mikepd&ProjectName=errsprite

Thank you for your help.
mike

**Comment by mikemm27** [comment]
Dear Jens,

unfortunately I have no final answer to my question and for this I would 
ask you again to consider the problem: IMHO I cannot find any reason
justifying the change of active sprite within the call of 
'run check of Sprite2' command and before its returning to the main program.

Thank you.
mike

**Comment by brianharvey** [comment]
I found your project way complicated, so I made a simpler case to check.  As Jens says, sprite-related blocks inside a ring refer to the sprite by which the ring was created, even when used with OF.  But sprite-related blocks inside a named custom block *do* apply to the sprite chosen by OF.  This seems inconsistent to me, but I'm guessing Jens won't agree, so I'm marking this issue "controversial misfeature" so we'll remember to discuss it.

**Comment by cycomachead** [comment]
> As Jens says, sprite-related blocks inside a ring refer to the sprite by which the ring was created, even when used with OF.

Wait, this sentence confuses me... I though the point of "zombification" was that the context of those blocks was the Sprite in the OF function (second arg).

**Comment by brianharvey** [comment]
Yes, that *is* the point, but sometimes it doesn't happen.  Looking again, the cases that do the surprising thing seem to have run inside (maybe indirectly) run:
![runrun](https://cloud.githubusercontent.com/assets/3885074/11461995/3a4accc8-96c2-11e5-80c3-83af1275d57c.png)
This moves Sprite, not Sprite2 (when clicked in Sprite's script area).

**Comment by cycomachead** [comment]
Ah! Now I see what you mean!

Yeah, that does seem confusing because the `RUN` is in the context of Sprite(2), but `foo` is Sprite. I guess, this is confusing but to me, perhaps seems correct. Essentially, it seems like all rings would then need to recursively "transfer" their context when passed in OF, which seems possible? Of course, the difference between using `RUN` and not using it, doesn't indicate about how the context changes...

So, yeah, I guess after all that, I'm not sure what's right.
